src/hotspot/cpu/aarch64/c2_MacroAssembler_aarch64.cpp

```diff
@@ -2557,3 +2557,18 @@ bool C2_MacroAssembler::in_scratch_emit_size() {
   }
   return MacroAssembler::in_scratch_emit_size();
 }
+
+void C2_MacroAssembler::load_nklass_compact_c2(Register dst, Register obj, int disp) {
+  // Note: Don't clobber obj anywhere in that method!
+
+  // The incoming address is pointing into obj-start + klass_offset_in_bytes. We need to extract
+  // obj-start, so that we can load from the object's mark-word instead. Usually the address
+  // comes as obj-start in obj and klass_offset_in_bytes in disp. However, sometimes C2
+  // combines decoding of a compressed oop and the load of the narrow Klass. When that happens,
+  // we get the heapBase in obj, and the narrowOop+klass_offset_in_bytes/sizeof(narrowOop) in index.
+  // When that happens, we need to lea the address into a single register, and subtract the
+  // klass_offset_in_bytes, to get the address of the mark-word.
```

Comment:
Parts of this comment are obsolete after commit 2c4a7877, please update the comment.